fuse_core/src/async_motion_model.cpp

```diff
@@ -142,42 +153,32 @@ void AsyncMotionModel::start()
 
 void AsyncMotionModel::stop()
 {
-  // Optimizers can sometimes ask the models to stop -> start without initializing.
-  // So we only fully wrap up if the context is no longer valid.
-  // Otherwise, we only reset the internal state by clearing the callback queue and calling onStop()
+  // Prefer to call onStop in executor's thread so downstream users don't have
+  // to worry about threads in ROS callbacks when there's only 1 thread.
```

Comment:
If we call `onStop` in the executor's thread, I think we should block with `wait()` in the calling thread, otherwise the calling thread might carry on, expecting an onStop to get called.

Alternatively, maybe we could return the future so the user can decide if they want to wait on it or not?

Reply:
Another reason for this is there is a possible race condition where if:
- Stop gets called, then start gets called

Then, in the multithreaded case, even though the callback queue is a queue, certain `onStop()` invocations might still be resolving when `onStart()` gets popped off in a different thread.

(That said, there are comments in the code that tell users to lock appropriately when using a multi-threaded executor, so we don't necessarily need to guard against this case.)

Reply:
> If we call onStop in the executor's thread, I think we should block with wait() in the calling thread, otherwise the calling thread might carry on, expecting onStop to have been called.

I think the code is already doing that. It gets the future for the `onStop` callback and waits on it in the calling thread. https://github.com/locusrobotics/fuse/pull/294/files#diff-327f5818028beb79e2c394523234fdeed441dd8d133721c7287273d05d58c892R162-R164

Reply:
Whoops.. I was looking at the graph callback wait() removal :man_facepalming: My bad!